src/CoreWCF.Primitives/src/CoreWCF/IdentityModel/Tokens/Saml2SecurityTokenHandler.cs

```diff
@@ -0,0 +1,200 @@
+using Microsoft.IdentityModel.Tokens.Saml2;
+using MSIdentityTokens = Microsoft.IdentityModel.Tokens;
+
+namespace CoreWCF.IdentityModel.Tokens
+{
+    /// <summary>
+    /// Creates SAML2 assertion-based security tokens
+    /// </summary>
+    public class Saml2SecurityTokenHandler : SecurityTokenHandler
+    {
+        /// <summary>
+        /// The key identifier value type for SAML 2.0 assertion IDs, as defined
+        /// by the OASIS Web Services Security SAML Token Profile 1.1. 
+        /// </summary>
+        public const string TokenProfile11ValueType = "http://docs.oasis-open.org/wss/oasis-wss-saml-token-profile-1.1#SAMLID";
+        private static readonly string[] s_tokenTypeIdentifiers = new string[] { SecurityTokenTypes.Saml2TokenProfile11, SecurityTokenTypes.OasisWssSaml2TokenProfile11 };
+        private SecurityTokenSerializer _keyInfoSerializer;
+        private readonly MSIdentityTokens.Saml2.Saml2SecurityTokenHandler _internalSaml2SecurityTokenHandler;
+        private readonly object _syncObject = new object();
+
+        /// <summary>
+        /// Creates an instance of <see cref="Saml2SecurityTokenHandler"/>
+        /// </summary>
+        public Saml2SecurityTokenHandler()
+           // : this(new Saml2SecurityTokenRequirement())
+        {
+            _internalSaml2SecurityTokenHandler = new MSIdentityTokens.Saml2.Saml2SecurityTokenHandler();
+        }
+
+        #region TokenValidation
+        /// <summary>
+        /// Returns value indicates if this handler can validate tokens of type
+        /// Saml2SecurityToken.
+        /// </summary>
+        public override bool CanValidateToken
+        {
+            get { return _internalSaml2SecurityTokenHandler.CanValidateToken; }
+        }
+
+        /// <summary>
+        /// Validates a <see cref="Saml2SecurityToken"/>.
+        /// </summary>
+        /// <param name="token">The <see cref="Saml2SecurityToken"/> to validate.</param>
+        /// <returns>The <see cref="ReadOnlyCollection{T}"/> of <see cref="ClaimsIdentity"/> representing the identities contained in the token.</returns>
+        /// <exception cref="ArgumentNullException">The parameter 'token' is null.</exception>
+        /// <exception cref="ArgumentException">The token is not assignable from <see cref="Saml2SecurityToken"/>.</exception>
+        /// <exception cref="InvalidOperationException">Configuration <see cref="SecurityTokenHandlerConfiguration"/>is null.</exception>
+        /// <exception cref="ArgumentException">Saml2SecurityToken.Assertion is null.</exception>
+        /// <exception cref="SecurityTokenValidationException">Thrown if Saml2SecurityToken.Assertion.SigningToken is null.</exception>
+        /// <exception cref="SecurityTokenValidationException">Thrown if the certificate associated with the token issuer does not pass validation.</exception>
+        public override ReadOnlyCollection<ClaimsIdentity> ValidateToken(SecurityToken token)
+        {
```

Comment:
There needs to be code here which creates a WindowsIdentity from a upn claim if configured. This is used for impersonation and delegation so can be an important scenario in a Windows domain environment. See the code [here](https://referencesource.microsoft.com/#System.IdentityModel/System/IdentityModel/Tokens/Saml2SecurityTokenHandler.cs,371)

Reply:
@mconnew the code looks like never been tested or never run ... we set it to null and try to find Upn, which will throw exceptions... so I am doubtful if it's ever run... hesitate to port the code... `[ClaimsIdentity](https://referencesource.microsoft.com/mscorlib/A.html#8437a543724aff81) claimsIdentity = null;
 
                if ([this](https://referencesource.microsoft.com/System.IdentityModel/System/IdentityModel/Tokens/Saml2SecurityTokenHandler.cs.html#140cc95ebaba4d7c).[samlSecurityTokenRequirement](https://referencesource.microsoft.com/System.IdentityModel/System/IdentityModel/Tokens/Saml2SecurityTokenHandler.cs.html#a0fb4157f95d63a5).[MapToWindows](https://referencesource.microsoft.com/System.IdentityModel/System/IdentityModel/Tokens/SamlSecurityTokenRequirement.cs.html#b7b93a539e2eb61d))
                {
                    // TFS: 153865, Microsoft WindowsIdentity does not set Authtype. I don't think that authtype should be set here anyway.
                    // The authtype will be S4U (kerberos) it doesn't really matter that the upn arrived in a SAML token.
                    claimsIdentity = [this](https://referencesource.microsoft.com/System.IdentityModel/System/IdentityModel/Tokens/Saml2SecurityTokenHandler.cs.html#140cc95ebaba4d7c).[CreateWindowsIdentity](https://referencesource.microsoft.com/System.IdentityModel/System/IdentityModel/Tokens/Saml2SecurityTokenHandler.cs.html#5ab264ef23a39a4b)([this](https://referencesource.microsoft.com/System.IdentityModel/System/IdentityModel/Tokens/Saml2SecurityTokenHandler.cs.html#140cc95ebaba4d7c).[FindUpn](https://referencesource.microsoft.com/System.IdentityModel/System/IdentityModel/Tokens/Saml2SecurityTokenHandler.cs.html#d108b35202c6e07e)(claimsIdentity));
 
                    // PARTIAL TRUST: will fail when adding claims, [AddClaims](https://referencesource.microsoft.com/mscorlib/A.html#0f18798b2eb03dd2) is SecurityCritical.
                    claimsIdentity.AddClaims([this](https://referencesource.microsoft.com/System.IdentityModel/System/IdentityModel/Tokens/Saml2SecurityTokenHandler.cs.html#140cc95ebaba4d7c).[CreateClaims](https://referencesource.microsoft.com/System.IdentityModel/System/IdentityModel/Tokens/Saml2SecurityTokenHandler.cs.html#d3bdd59d5052ea7c)(samlToken).[Claims](https://referencesource.microsoft.com/mscorlib/A.html#5550b64fc5cd4474));
                }
                else
                {
                    claimsIdentity = [this](https://referencesource.microsoft.com/System.IdentityModel/System/IdentityModel/Tokens/Saml2SecurityTokenHandler.cs.html#140cc95ebaba4d7c).[CreateClaims](https://referencesource.microsoft.com/System.IdentityModel/System/IdentityModel/Tokens/Saml2SecurityTokenHandler.cs.html#d3bdd59d5052ea7c)(samlToken);
                }`